File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.py
def solution(n, words):
    answer = [0, 0]
    
    prev_words: set[str] = set()
    end_word: str = words[0][-1]
    prev_words.add(words[0])
    for i in range(1, len(words)):
        word: str = words[i]
        if word in prev_words or not word.startswith(end_word):
            answer[0] = i % n + 1
            answer[1] = i // n + 1
            break
        prev_words.add(word)
        end_word = word[-1]

    return answer

File: 프로그래머스/2/12981. 영어 끝말잇기/영어 끝말잇기.java
import java.util.*;

class Solution {
    public int[] solution(int n, String[] words) {
        int[] answer = new int[2];
        Set<String> usedWords = new HashSet<>();
        String prevWord = words[0];
        usedWords.add(prevWord);
        
        boolean success = true;
        int i;
        for (i = 1; i < words.length; i++) {
            String word = words[i];
            if (
                usedWords.contains(word) || word.charAt(0) != prevWord.charAt(prevWord.length() - 1)
            ) {
                success = false;
                break;
            }
            usedWords.add(word);
            prevWord = word;
        }
        
        if (success) {
            return answer;
        }
        answer[0] = (i % n) + 1;
        answer[1] = i / n + 1;
        return answer;
    }
}